src/region_binary_search.py

```diff
@@ -0,0 +1,40 @@
+from bed import BedLine
+
+def extract_region(features: list[BedLine],
+                   start: int, end: int) -> list[BedLine]:
+    """Extract region chrom[start:end] and write it to outfile."""
+    index_start = binary_search_region_start(features, start)
+    filtered = list()
+    for i in range(index_start, len(features)):
+        feat = features[i]
+        if feat.chrom_start >= end:
+            break
+        filtered.append(feat)
+    return filtered
+
+def is_overlapping(x: tuple[int,int], y: tuple[int,int]) -> bool:
+    # Credits to Thomas Maillund :)
+    return max(x[0], y[0]) < min(x[1], y[1])
+
+def is_feature_in(feat: BedLine, start: int, end: int):
+    interval = (feat.chrom_start, feat.chrom_end)
+    return is_overlapping(interval, (start, end))
+
+def binary_search_region_start(x: list[BedLine], start: int, low = None, high = None):
+    low, high = 0, len(x)
+    if not x:
+        return 0
+    while True:
+        mid = (high + low) // 2
+        y = x[mid].chrom_start
+        if y == start:
+            return mid
```

Comment:
All you know here is that `mid` is an index where you have a match. You do not know that it is the first. You risk missing some that were before the midpoint.

You can simplify this search a lot if you use the following invariant: All indices below `low`, `range(0,low)`, should have a start point before `start`. All indices above `high`, `range(high,len(x))`, should have a start point after or equal to `start`. That allows you to move `low` up to `mid+1` if index `mid` is _smaller_ than start (because all below `low` must be smaller, so only if `mid` is _smaller_ can we move `low` up. In any other case, `mid` is greater or equal to `start`, and then you are allowed to move `high` down.

If you continue doing this until either `x[low]` is equal to `start` or until `low == high`, you have your index in `low`.
The first case because we only move `low` up to `mid+1` when `mid` is too small, so if `x[low]` matches `start`, then the previous index didn't, so we have the smallest. The second because we have searched for `start` until either we didn't find it so we now have an empty interval, or because we had `high = low + 1`, then got the midpoint `mid = low` that was too small, and then moved up (which puts us in case 1 again).